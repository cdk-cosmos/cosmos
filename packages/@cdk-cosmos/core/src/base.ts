import { IConstruct, CfnResource } from '@aws-cdk/core';
import { Construct, ConstructNode } from '@aws-cdk/core/lib/construct-compat';
import { NetworkBuilder } from '@aws-cdk/aws-ec2/lib/network-util';
import { Stack, StackProps } from '@aws-cdk/core/lib/stack';

export const COSMOS_PARTITION = 'COSMOS_PARTITION';
export const COSMOS_VERSION = 'COSMOS_VERSION';
export const COSMOS_NETWORK_BUILDER = 'COSMOS_NETWORK_BUILDER';
export const AWS_ENV = 'AWS_ENV';

export const PATTERN = {
  SINGLETON_COSMOS: '{Partition}{Resource}+',
  SINGLETON_GALAXY: '{Partition}{Galaxy}{Resource}+',
  SINGLETON_SOLAR_SYSTEM: '${Partition}{Galaxy}{SolarSystem}{Resource}+',
  STACK: '{Partition}{Cosmos}{Galaxy}?{SolarSystem}?{Version}?{Type}',
  COSMOS: '{Partition}{Cosmos}{Galaxy}?{SolarSystem}?{Resource}*{Version}?',
  RESOURCE: '{Resource}+{Version}?',
  // SHORT_SOLAR_SYSTEM: '{Partition}{Cosmos}{SolarSystem}{Resource}+',
  // DOCKER_TAG: '{Cosmos}/{Resource}+',
  // LOG_GROUP: '{Partition}/{Cosmos}/{SolarSystem}/{Resource}+',
};

declare module '@aws-cdk/core/lib/construct-compat' {
  interface ConstructNode {
    id: string;
    type: string;
    pattern: string;
  }

  interface Construct {
    generateId(id: string, delimiter?: string, pattern?: string, type?: string): string;
    singletonId(id: string, delimiter?: string, type?: string): string;
  }
}

ConstructNode.prototype.type = 'Resource';

Construct.prototype.generateId = function(id, delimiter, pattern, type): string {
  return generateNodeId({
    scope: this,
    pattern,
    id,
    type,
    delimiter,
  });
};

Construct.prototype.singletonId = function(id, delimiter, type): string {
  return singletonNodeId({
    scope: this,
    id,
    type,
    delimiter,
  });
};

export interface IBase extends IConstruct {
  NetworkBuilder?: NetworkBuilder;
}

export interface BaseConstructProps {
  type: string;
  cidr?: string;
}

export class BaseConstruct extends Construct implements IBase {
  public readonly networkBuilder?: NetworkBuilder;

  constructor(scope: Construct, id: string, props: BaseConstructProps) {
    super(scope, id);
    const { type, cidr } = props;

    this.node.type = type;
    if (cidr) this.networkBuilder = new NetworkBuilder(cidr);
    if (this.networkBuilder) this.node.setContext(COSMOS_NETWORK_BUILDER, this.networkBuilder);
    else this.networkBuilder = this.node.tryGetContext(COSMOS_NETWORK_BUILDER);
  }
}

export interface BaseStackProps extends StackProps {
  partition?: string;
  version?: string;
  type: string;
  disableCosmosNaming?: boolean;
}

export class BaseStack<T extends IBase> extends Stack {
  private readonly disableCosmosNaming: boolean;
  protected _resource: T;
  public get resource(): T {
    return this._resource;
  }

  constructor(scope: Construct, id: string, props: BaseStackProps) {
    const { partition, version, type, disableCosmosNaming = false, env } = props;

    super(scope, id, {
      stackName: generateNodeId({ scope, id, type, partition, version, pattern: PATTERN.STACK }),
      env: scope.node.tryGetContext(AWS_ENV),
      ...props,
    });

    this.node.type = type;
    if (partition) this.node.setContext(COSMOS_PARTITION, partition);
    if (version) this.node.setContext(COSMOS_VERSION, version);
    this.disableCosmosNaming = disableCosmosNaming;
    if (env) this.node.setContext(AWS_ENV, env);
  }

  public allocateLogicalId(scope: CfnResource): string {
    if (this.disableCosmosNaming) return super.allocateLogicalId(scope);

    const id = generateId({
      scopes: getScopes(scope),
      pattern: getPattern(scope) || PATTERN.RESOURCE,
      partition: scope.node.tryGetContext(COSMOS_PARTITION),
      version: scope.node.tryGetContext(COSMOS_VERSION),
    });

    return removeNonAlphanumeric(id);
  }
}

export function singletonNodeId(props: { scope: IConstruct; id: string; type?: string; delimiter?: string }): string {
  const { scope, id, type, delimiter } = props;
  let pattern: string;
  switch (scope.node.type) {
    case 'Cosmos':
      pattern = PATTERN.SINGLETON_COSMOS;
      break;
    case 'Galaxy':
      pattern = PATTERN.SINGLETON_GALAXY;
      break;
    case 'SolarSystem':
      pattern = PATTERN.SINGLETON_SOLAR_SYSTEM;
      break;
    default:
      throw new Error(`Singleton Pattern could not be found for ${scope.node.id}`);
  }
  return generateNodeId({
    scope,
    pattern,
    id,
    type,
    delimiter,
  });
}

export function generateNodeId(props: {
  scope: IConstruct;
  id: string;
  pattern?: string;
  type?: string;
  partition?: string;
  version?: string;
  delimiter?: string;
}): string {
  const { scope, id, pattern, type = 'Resource', partition, version, delimiter } = props;
  return generateId({
    scopes: [...getScopes(scope), { key: type, value: id }, { key: 'Type', value: type }],
    pattern: pattern || PATTERN.COSMOS,
    partition: partition || scope.node.tryGetContext(COSMOS_PARTITION),
    version: version || scope.node.tryGetContext(COSMOS_VERSION),
    delimiter,
  });
}

interface IScope {
  key: string;
  value: string;
}

interface GenerateLogicalIdProps {
  scopes: IScope[];
  pattern: string;
  partition?: string;
  version?: string;
  delimiter?: string;
}

export function generateId(props: GenerateLogicalIdProps): string {
  const { scopes, pattern, partition, version, delimiter = '' } = props;

  if (partition) scopes.push({ key: 'Partition', value: partition });
  if (version) scopes.push({ key: 'Version', value: version });

  const selectedIds = selectScoped(scopes.filter(removeHidden), pattern)
    .map(x => x.value)
    .reduce(removeDupes, []);

  return selectedIds.join(delimiter).slice(0, 240);
}

function selectScoped(scopes: IScope[], pattern: string): IScope[] {
  const regex = /{(\w+)}([*+?])?/gm;
  const results: IScope[] = [];
  let segment: RegExpExecArray | null;
  while ((segment = regex.exec(pattern)) !== null) {
    const key: string = segment[1];
    const selector: string | undefined = segment[2];

    switch (selector) {
      case '*': {
        const selected = scopes.filter(x => x.key === key);
        if (selected.length) results.push(...selected);
        break;
      }
      case '+': {
        if (!scopes.some(x => x.key === key)) throw new Error(`No ${key} Scope found`);
        results.push(...scopes.filter(x => x.key === key));
        break;
      }
      case '?': {
        const selected = scopes.filter(x => x.key === key).pop();
        if (selected) results.push(selected);
        break;
      }
      default: {
        const selected = scopes.filter(x => x.key === key).pop();
        if (!selected) throw new Error(`No ${key} Scope found.`);
        results.push(selected);
      }
    }
  }

  return results;
}

function getScopes(scope: IConstruct): IScope[] {
  const scopes = scope.node.scopes
    .filter(x => x.node.id && x.node.type)
    .map(x => ({
      key: x.node.type,
      value: x.node.id,
    }));

  if (scope.node.type) scopes.push({ key: 'Type', value: scope.node.type });

  return scopes;
}

function getPattern(scope: IConstruct): string | undefined {
  return scope.node.scopes
    .map(x => x.node.pattern)
    .filter(x => x)
    .pop();
}

function removeDupes(result: string[], value: string): string[] {
  if (result.length === 0 || !result[result.length - 1].endsWith(value)) {
    result.push(value);
  }

  return result;
}

function removeHidden(scope: IScope): boolean {
  if (scope.value === 'Default' || scope.value === 'Resource') return false;
  return true;
}

function removeNonAlphanumeric(s: string): string {
  return s.replace(/[^A-Za-z0-9]/g, '');
}

// ---- ---- ---- ----

// /**
//  * Resources with this ID are hidden from humans
//  *
//  * They do not appear in the human-readable part of the logical ID,
//  * but they are included in the hash calculation.
//  */
// const HIDDEN_FROM_HUMAN_ID = 'Resource';

// /**
//  * Resources with this ID are complete hidden from the logical ID calculation.
//  */
// const HIDDEN_ID = 'Default';

// const PATH_SEP = '/';

// const HASH_LEN = 8;
// const MAX_HUMAN_LEN = 240; // max ID len is 255
// const MAX_ID_LEN = 255;

// /**
//  * Calculates a unique ID for a set of textual components.
//  *
//  * This is done by calculating a hash on the full path and using it as a suffix
//  * of a length-limited "human" rendition of the path components.
//  *
//  * @param components The path components
//  * @returns a unique alpha-numeric identifier with a maximum length of 255
//  */
// export function makeUniqueId(components: string[]) {
//   components = components.filter(x => x !== HIDDEN_ID);

//   if (components.length === 0) {
//     throw new Error('Unable to calculate a unique id for an empty set of components');
//   }

//   // Lazy require in order to break a module dependency cycle
//   const unresolvedTokens = components.filter(c => unresolved(c));
//   if (unresolvedTokens.length > 0) {
//     throw new Error(`ID components may not include unresolved tokens: ${unresolvedTokens.join(',')}`);
//   }

//   // top-level resources will simply use the `name` as-is in order to support
//   // transparent migration of cloudformation templates to the CDK without the
//   // need to rename all resources.
//   if (components.length === 1) {
//     // we filter out non-alpha characters but that is actually a bad idea
//     // because it could create conflicts ("A-B" and "AB" will render the same
//     // logical ID). sadly, changing it in the 1.x version line is impossible
//     // because it will be a breaking change. we should consider for v2.0.
//     // https://github.com/aws/aws-cdk/issues/6421
//     const candidate = removeNonAlphanumeric(components[0]);

//     // if our candidate is short enough, use it as is. otherwise, fall back to
//     // the normal mode.
//     if (candidate.length <= MAX_ID_LEN) {
//       return candidate;
//     }
//   }

//   const hash = pathHash(components);
//   const human = removeDupes(components)
//     .filter(x => x !== HIDDEN_FROM_HUMAN_ID)
//     .map(removeNonAlphanumeric)
//     .join('')
//     .slice(0, MAX_HUMAN_LEN);

//   return human + hash;
// }

// /**
//  * Take a hash of the given path.
//  *
//  * The hash is limited in size.
//  */
// function pathHash(path: string[]): string {
//   const md5 = crypto
//     .createHash('md5')
//     .update(path.join(PATH_SEP))
//     .digest('hex');
//   return md5.slice(0, HASH_LEN).toUpperCase();
// }

// /**
//  * Removes all non-alphanumeric characters in a string.
//  */
// function removeNonAlphanumeric(s: string) {
//   return s.replace(/[^A-Za-z0-9]/g, '');
// }

/**
 * Remove duplicate "terms" from the path list
 *
 * If the previous path component name ends with this component name, skip the
 * current component.
 */
// function removeDupes(path: string[]): string[] {
//   const ret = new Array<string>();

//   for (const component of path) {
//     if (ret.length === 0 || !ret[ret.length - 1].endsWith(component)) {
//       ret.push(component);
//     }
//   }

//   return ret;
// }
