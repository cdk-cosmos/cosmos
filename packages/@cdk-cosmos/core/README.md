# Cosmos Core Package

This Library contains the base components for the CDK Cosmos. It defiles the Core Classes and the Extension Classes for the Core.
All other @cdk-cosmos packages will build on top of this core concept.

## Basic Understanding

Cosmos is broken down into 3 scopes (Bubbles)

- Cosmos: Top level scope of the entire infrastructure, mostly singleton shared resources.
- Galaxy: Account level scope that defines infrastructure for an aws account.
- SolarSystem: Environment level scope that defines infrastructure for an environment with in an aws account (Dev, Tst, Prd).

## Cosmos Core

The Cosmos core contains infrastructure that will be shared by consumers building apps (extensions).

The Core Cosmos is composed of:

- CosmosStack
  - TLD Zone
  - CDK Repo - for the core cdk code
  - CDK Master Role - for builds of the cdk deployments
- GalaxyStack
  - If Cross Account, create Cross Account CDK Role for master to assume
- SolarSystemStack
  - VPC (Created on the Galaxy scope, so that it can be shared by many SolarSystems in the same Galaxy)
  - Zone
- EcsSolarSystemStack
  - Cluster
  - Alb
  - HttpListener
- CiCdSolarSystemStack
  - VPC
  - Zone
  - Cluster
  - Alb
  - HttpListener
  - CDK Pipeline + Code Build for Deployment of the core

### Example

```js
const cosmos = new CosmosStack(app, 'Demo', {
  tld: 'cosmos.com',
  cidr: '10.0.0.0/22',
});

const mgtGalaxy = new GalaxyStack(cosmos, 'Mgt');

const ciCd = new CiCdSolarSystemStack(mgtGalaxy);

const dev = new EcsSolarSystemStack(mgtGalaxy, 'Dev');
```

## Cosmos Extensions

The Extensions are a way of consuming and adding resources in an isolated way. Each Extension has a Portal Property that is used to consume that core infrastructure for that same bubble.

The Extension of the core are:

- CosmosExtensionStack
  - Imports Core Cosmos
  - CDK Repo - for the app cdk code
- GalaxyExtensionStack
  - Imports Core Galaxy
- SolarSystemExtensionStack
  - Imports Core SolarSystem
- EcsSolarSystemExtensionStack
  - Imports Core EcsSolarSystem
- CiCdSolarSystemExtensionStack
  - Imports Core CiCdSolarSystem
  - CDK Pipeline + Code Build for Deployment of the app

### Example

```js
const cosmos = new CosmosExtensionStack(app, 'Demo');

const mgtGalaxy = new GalaxyExtensionStack(cosmos, 'Mgt');

const ciCd = new CiCdSolarSystemExtensionStack(mgtGalaxy);

const dev = new EcsSolarSystemExtensionStack(mgtGalaxy, 'Dev');
```

What is included in the above is just enough for the extension (App) to host its own code and deploy it self. From here we expect consumer to add their own infrastructure to extends from the core, into the respective bubbles.

## Further Use

For more extensive understanding and usage, please look at the docs, aka the [law of the cosmos](https://github.com/cdk-cosmos/law)

Deploy CDKToolKit `npx cdk deploy --app="node_modules/@cosmos-building-blocks/common/lib/cdk-tool-kit-app.js"`