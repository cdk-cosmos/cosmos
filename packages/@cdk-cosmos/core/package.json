{
  "name": "@cdk-cosmos/core",
  "version": "0.2.0",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib/*",
    "package.json",
    "README.md"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cdk-cosmos/cosmos.git",
    "directory": "packages/@cdk-cosmos/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "test": "jest",
    "lint": "eslint . --ext .js,.ts",
    "format": "prettier --write **/*.ts !**/*.d.ts",
    "clean": "rm -rf lib"
  },
  "dependencies": {
    "@cosmos-building-blocks/common": "^0.2.0"
  },
  "peerDependencies": {
    "@aws-cdk/aws-cloudformation": "^1.23.0",
    "@aws-cdk/aws-codebuild": "^1.23.0",
    "@aws-cdk/aws-codecommit": "^1.23.0",
    "@aws-cdk/aws-codepipeline": "^1.23.0",
    "@aws-cdk/aws-codepipeline-actions": "^1.23.0",
    "@aws-cdk/aws-ec2": "^1.23.0",
    "@aws-cdk/aws-ecr": "^1.23.0",
    "@aws-cdk/aws-ecs": "^1.23.0",
    "@aws-cdk/aws-elasticloadbalancingv2": "^1.23.0",
    "@aws-cdk/aws-elasticloadbalancingv2-targets": "^1.23.0",
    "@aws-cdk/aws-iam": "^1.23.0",
    "@aws-cdk/aws-lambda": "^1.23.0",
    "@aws-cdk/aws-route53": "^1.23.0",
    "@aws-cdk/core": "^1.23.0"
  },
  "gitHead": "173c051038a3acdaa93d6a37d42b6b0546ffd567"
}
