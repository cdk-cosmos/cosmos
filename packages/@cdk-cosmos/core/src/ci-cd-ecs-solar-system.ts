import { Project } from '@aws-cdk/aws-codebuild';
import {
  IEcsSolarSystemCore,
  EcsSolarSystemCoreStack,
  EcsSolarSystemCoreProps,
  ImportedEcsSolarSystemCore,
  EcsSolarSystemExtensionStack,
  IEcsSolarSystemExtension,
} from './ecs-solar-system';
import { IGalaxyExtension, IGalaxyCore } from './galaxy';
import { CdkPipeline } from './components/cdk-pipeline';
import {
  ICiCdSolarSystemCore,
  ICiCdSolarSystemExtension,
  CiCdSolarSystemCoreStackProps,
  CiCdSolarSystemExtensionStackProps,
  CosmosCdkPipeline,
} from './ci-cd-solar-system';

export interface ICiCdEcsSolarSystemCore extends ICiCdSolarSystemCore, IEcsSolarSystemCore {}

export interface CiCdEcsSolarSystemCoreStackProps extends CiCdSolarSystemCoreStackProps, EcsSolarSystemCoreProps {}

export class CiCdEcsSolarSystemCoreStack extends EcsSolarSystemCoreStack implements ICiCdEcsSolarSystemCore {
  readonly deployPipeline: CdkPipeline;
  readonly deployProject: Project;

  constructor(galaxy: IGalaxyCore, props?: CiCdEcsSolarSystemCoreStackProps) {
    super(galaxy, 'CiCd', {
      description: 'Cosmos CiCdEcsSolarSystem: Resources dependant on Ci & Cd, like CodePipelines and CodeDeployments.',
      ...props,
    });

    const { cdkPipelineProps } = props || {};

    this.deployPipeline = new CosmosCdkPipeline(this, 'CdkPipeline', cdkPipelineProps);
    this.deployProject = this.deployPipeline.Deploy;
  }
}

export interface ICiCdEcsSolarSystemExtension extends ICiCdSolarSystemExtension, IEcsSolarSystemExtension {
  portal: ICiCdEcsSolarSystemCore;
}

export interface CiCdEcsSolarSystemExtensionStackProps extends CiCdSolarSystemExtensionStackProps {}

export class CiCdEcsSolarSystemExtensionStack extends EcsSolarSystemExtensionStack
  implements ICiCdEcsSolarSystemExtension {
  readonly portal: ICiCdEcsSolarSystemCore;
  readonly deployPipeline: CdkPipeline;
  readonly deployProject: Project;

  constructor(galaxy: IGalaxyExtension, props?: CiCdEcsSolarSystemExtensionStackProps) {
    super(galaxy, 'CiCd', {
      description:
        'Cosmos CiCdEcsSolarSystem Extension: App resources dependant on Ci & Cd, like CodePipelines and CodeDeployments.',
      ...props,
    });

    const { cdkPipelineProps } = props || {};

    this.deployPipeline = new CosmosCdkPipeline(this, 'CdkPipeline', cdkPipelineProps);
    this.deployProject = this.deployPipeline.Deploy;
  }

  protected getPortal(props?: CiCdSolarSystemExtensionStackProps): ICiCdEcsSolarSystemCore {
    const galaxy = this.node.scope as IGalaxyExtension;
    return new ImportedEcsSolarSystemCore(galaxy.portal, this.node.id, props?.portalProps);
  }
}
