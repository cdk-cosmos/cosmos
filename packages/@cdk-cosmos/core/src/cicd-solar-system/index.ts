export {
  ICiCdSolarSystemCore,
  CiCdSolarSystemCoreStack,
  CiCdEcsSolarSystemCoreStack,
  CiCdSolarSystemCoreStackProps,
} from './cicd-solar-system-core-stack';
export {
  ICiCdSolarSystemExtension,
  CiCdSolarSystemExtensionStack,
  CiCdEcsSolarSystemExtensionStack,
  CiCdSolarSystemExtensionStackProps,
} from './cicd-solar-system-extension-stack';
