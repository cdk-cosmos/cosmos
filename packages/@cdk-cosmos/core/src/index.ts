export {
  Bubble,
  Cosmos,
  Galaxy,
  SolarSystem,
  EcsSolarSystem,
  CiCdSolarSystem,
  Extension,
  CosmosExtension,
  GalaxyExtension,
  SolarSystemExtension,
  EcsSolarSystemExtension,
  CiCdSolarSystemExtension,
} from './interfaces';

export {
  RemoteZone,
  RemoteVpc,
  RemoteCluster,
  RemoteAlb,
  RemoteApplicationListener,
  RemoteCodeRepo,
  RemoteBuildProject,
} from './remote';

export { CosmosStack, CosmosStackProps, ImportedCosmos, CosmosExtensionStack } from './cosmos';

export { GalaxyStack, GalaxyStackProps, ImportedGalaxy, GalaxyExtensionStack } from './galaxy';

export { SolarSystemStack, SolarSystemProps, ImportedSolarSystem, SolarSystemExtensionStack } from './solar-system';

export {
  EcsSolarSystemStack,
  EcsSolarSystemProps,
  ImportedEcsSolarSystem,
  EcsSolarSystemExtensionStack,
} from './ecs-solar-system';

export {
  CiCdSolarSystemStack,
  CiCdStackProps,
  ImportedCiCdSolarSystem,
  CiCdSolarSystemExtensionStack,
} from './ci-cd-solar-system';

export { CdkPipeline, CdkPipelineProps } from './cdk-pipeline';

export { RESOLVE, PATTERN, _RESOLVE, ResolveParams } from './pattern';
