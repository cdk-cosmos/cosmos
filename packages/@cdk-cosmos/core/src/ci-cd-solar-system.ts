import { Construct, StackProps } from '@aws-cdk/core';

import { Project } from '@aws-cdk/aws-codebuild';
import { Role } from '@aws-cdk/aws-iam';
import {
  PATTERN,
  Galaxy,
  EcsSolarSystem,
  CiCdSolarSystem,
  GalaxyExtension,
  EcsSolarSystemStack,
  EcsSolarSystemProps,
  ImportedEcsSolarSystem,
  EcsSolarSystemExtensionStack,
  CiCdSolarSystemExtension,
  CdkPipeline,
  CdkPipelineProps,
} from '.';

export interface CiCdStackProps extends EcsSolarSystemProps {
  cdkPipelineProps?: Partial<CdkPipelineProps>;
}

export class CiCdSolarSystemStack extends EcsSolarSystemStack implements CiCdSolarSystem {
  readonly CdkDeploy: Project;

  constructor(galaxy: Galaxy, props?: CiCdStackProps) {
    super(galaxy, 'CiCd', props);

    const { cdkPipelineProps = {} } = props || {};
    const { CdkRepo, CdkMasterRoleStaticArn } = this.Galaxy.Cosmos;

    const pipeline = new CdkPipeline(this, 'CdkPipeline', {
      name: this.RESOLVE(PATTERN.SINGLETON_COSMOS, 'Cdk-Pipeline'),
      cdkRepo: CdkRepo,
      deployRole: Role.fromRoleArn(this, 'CdkMasterRole', CdkMasterRoleStaticArn, { mutable: false }),
      // deployVpc: this.Vpc,
      deployStacks: [this.RESOLVE(PATTERN.COSMOS, '*')],
      ...cdkPipelineProps,
    });
    this.CdkDeploy = pipeline.Deploy;

    // RemoteBuildProject.export(`Core${this.Account.Name}${this.Name}`, this.CdkDeploy);
  }
}

export class ImportedCiCdSolarSystem extends ImportedEcsSolarSystem implements CiCdSolarSystem {
  // readonly CdkDeploy: IProject;

  constructor(scope: Construct, galaxy: Galaxy) {
    super(scope, galaxy, 'CiCd');

    // this.CdkDeploy = RemoteBuildProject.import(this, `Core${this.Galaxy.Name}${this.Name}`, 'CdkPipelineDeploy');
  }
}

export class CiCdSolarSystemExtensionStack extends EcsSolarSystemExtensionStack implements CiCdSolarSystemExtension {
  readonly Portal: EcsSolarSystem;
  readonly DeployPipeline: CdkPipeline;
  readonly DeployProject: Project;

  constructor(galaxy: GalaxyExtension, props?: StackProps) {
    super(galaxy, 'CiCd', props);

    this.node.tryRemoveChild(this.Portal.node.id);
    this.Portal = new ImportedCiCdSolarSystem(this, this.Galaxy.Portal);

    this.DeployPipeline = new CdkPipeline(this, 'CdkPipeline', {
      name: `App-${this.Galaxy.Cosmos.Name}-Cdk-Pipeline`,
      cdkRepo: this.Galaxy.Cosmos.CdkRepo,
      deployRole: Role.fromRoleArn(this, 'CdkMasterRole', this.Galaxy.Cosmos.Portal.CdkMasterRoleStaticArn, {
        mutable: false,
      }),
      deployStacks: [this.RESOLVE(PATTERN.COSMOS, '*')],
    });
    this.DeployProject = this.DeployPipeline.Deploy;
  }
}
