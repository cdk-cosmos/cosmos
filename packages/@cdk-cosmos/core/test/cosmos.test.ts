import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { synthesizeStacks, toHaveResourceId, toHaveResourceCount } from '../../../../src/test';
import { CosmosCoreStack, CosmosExtensionStack } from '../src';

const app = new App();
const cosmos = new CosmosCoreStack(app, 'Cos', { tld: 'cos.com', cidr: '10.0.0.0/24' });
const cosmosExtension = new CosmosExtensionStack(app, 'Test');
const [cosmosStack, cosmosExtensionStack] = synthesizeStacks(cosmos, cosmosExtension);

describe('Cosmos', () => {
  test('should be a cosmos', () => {
    expect(cosmosStack.name).toEqual('CoreCosCosmos');
    expect(cosmosStack).toHaveOutput({ exportName: 'CoreId', outputValue: 'Cos' });
    expect(cosmosStack).toHaveOutput({ exportName: 'CoreLibVersion' });
    toHaveResourceCount(cosmosStack, 4);
  });

  test('should have a RootZone', () => {
    expect(cosmosStack).toHaveResource('AWS::Route53::HostedZone', { Name: 'cos.com.' });
    expect(cosmosStack).toHaveOutput({ exportName: 'CoreRootZoneId' });
    expect(cosmosStack).toHaveOutput({ exportName: 'CoreRootZoneName' });
    toHaveResourceId(cosmosStack, 'RootZone');
  });

  test('should have a CdkRepo', () => {
    expect(cosmosStack).toHaveResource('AWS::CodeCommit::Repository', { RepositoryName: 'core-cos-cdk-repo' });
    expect(cosmosStack).toHaveOutput({ exportName: 'CoreCdkRepoName' });
    toHaveResourceId(cosmosStack, 'CdkRepo');
  });

  test('should have a CdkMasterRole', () => {
    expect(cosmosStack).toHaveResource('AWS::IAM::Role', { RoleName: 'CoreCdkMasterRole' });
    toHaveResourceId(cosmosStack, 'CdkMasterRole');
  });

  test('should have cidr range', () => {
    expect(cosmos.networkBuilder?.addSubnet(28)).toEqual('10.0.0.0/28');
    expect(cosmos.networkBuilder?.addSubnet(28)).toEqual('10.0.0.16/28');
  });

  test('should match snapshot', () => {
    expect(cosmosStack.template).toMatchSnapshot({
      Outputs: {
        CoreLibVersion: {
          Value: expect.any(String),
        },
      },
    });
  });
});

describe('Cosmos Extension', () => {
  test('should be a cosmos extension', () => {
    expect(cosmosExtensionStack.name).toEqual('AppTestCosmos');
    toHaveResourceCount(cosmosExtensionStack, 1);
  });

  test('should have a CdkRepo', () => {
    expect(cosmosExtensionStack).toHaveResource('AWS::CodeCommit::Repository', { RepositoryName: 'app-test-cdk-repo' });
    toHaveResourceId(cosmosExtensionStack, 'CdkRepo');
  });

  //TODO: Check if imports align

  test('should match snapshot', () => {
    expect(cosmosExtensionStack.template).toMatchSnapshot();
  });
});
