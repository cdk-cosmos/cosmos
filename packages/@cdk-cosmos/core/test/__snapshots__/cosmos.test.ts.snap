// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cosmos Extension should match snapshot 1`] = `
Object {
  "Description": "App Singleton Resources for the Cosmos like CdkRepo and EcrRepo.",
  "Resources": Object {
    "CdkRepo9606A710": Object {
      "Properties": Object {
        "RepositoryDescription": "App CDK Repo for Test Cosmos.",
        "RepositoryName": "app-test-cdk-repo",
      },
      "Type": "AWS::CodeCommit::Repository",
    },
  },
}
`;

exports[`Cosmos should match snapshot 1`] = `
Object {
  "Description": "Singleton resources for the Cosmos, like RootZone, CdkRepo and CdkMasterRole",
  "Outputs": Object {
    "CdkRepoRepoNameEBDD9CD6": Object {
      "Export": Object {
        "Name": "Core-CdkRepo-Name",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "CdkRepo9606A710",
          "Name",
        ],
      },
    },
    "CosmosName": Object {
      "Export": Object {
        "Name": "Core-Name",
      },
      "Value": "Cos",
    },
    "CosmosVersion": Object {
      "Export": Object {
        "Name": "Core-Version",
      },
      "Value": Any<String>,
    },
    "CrossAccountExportsFnFunctionArn4D0D888C": Object {
      "Export": Object {
        "Name": "Core-CrossAccountExportsFn-Arn",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "CrossAccountExportsFnBB7349E9",
          "Arn",
        ],
      },
    },
    "RootZoneNameServersC9CDB123": Object {
      "Export": Object {
        "Name": "Core-RootZone-NameServers",
      },
      "Value": Object {
        "Fn::Join": Array [
          ",",
          Object {
            "Fn::GetAtt": Array [
              "RootZone831A5F27",
              "NameServers",
            ],
          },
        ],
      },
    },
    "RootZoneZoneId57AFCBDD": Object {
      "Export": Object {
        "Name": "Core-RootZone-Id",
      },
      "Value": Object {
        "Ref": "RootZone831A5F27",
      },
    },
    "RootZoneZoneNameD5248AD9": Object {
      "Export": Object {
        "Name": "Core-RootZone-Name",
      },
      "Value": "cos.com",
    },
  },
  "Resources": Object {
    "CdkMasterRole2FE9B317": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "codebuild.amazonaws.com",
                  "codepipeline.amazonaws.com",
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
        "RoleName": "Core-CdkMaster-Role",
      },
      "Type": "AWS::IAM::Role",
    },
    "CdkRepo9606A710": Object {
      "Properties": Object {
        "RepositoryDescription": "Core CDK Repo for Cos Cosmos.",
        "RepositoryName": "core-cdk-repo",
      },
      "Type": "AWS::CodeCommit::Repository",
    },
    "CrossAccountExportsFnBB7349E9": Object {
      "DependsOn": Array [
        "CdkMasterRole2FE9B317",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "mock-s3BucketName",
          "S3Key": "mock-s3ObjectKey",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CdkMasterRole2FE9B317",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "RootZone831A5F27": Object {
      "Properties": Object {
        "HostedZoneConfig": Object {
          "Comment": "Core TLD Root Zone for Cos Cosmos.",
        },
        "Name": "cos.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
  },
}
`;
